### PR TITLE
Adds integration tests

### PR DESCRIPTION
**Pull request by claudiubelu**
The integration tests will build the charm from scratch, and deploy it in a currently bootstrapped Juju Controller in a new test model.

The integration tests will also deploy the ``postgresql-k8s`` and ``nginx-ingress-integrator`` charms and relates them to the ``finos-waltz-k8s`` charm. The tests are asserting that Waltz can be reached through its IP and through its default ``service-hostname``.

Partially-fixes: https://github.com/pedroleaoc/waltz-integration-juju/issues/8

### CODE REVIEW

**Comment by claudiubelu** [comment]
Sample run:

```
tox -e integration
integration installed: asttokens==2.0.5,attrs==21.4.0,backcall==0.2.0,bcrypt==3.2.0,cachetools==5.0.0,certifi==2021.10.8,cffi==1.15.0,charset-normalizer==2.0.12,cryptography==36.0.1,decorator==5.1.1,executing==0.8.3,google-auth==2.6.0,idna==3.3,iniconfig==1.1.1,ipdb==0.13.9,ipython==8.1.0,jedi==0.18.1,Jinja2==3.0.3,juju==2.9.7,jujubundlelib==0.5.7,kubernetes==23.3.0,macaroonbakery==1.3.1,MarkupSafe==2.1.0,matplotlib-inline==0.1.3,mypy-extensions==0.4.3,oauthlib==3.2.0,ops==1.3.0,ops-lib-pgsql==1.4,packaging==21.3,paramiko==2.9.2,parso==0.8.3,pexpect==4.8.0,pgconnstr==1.0.1,pickleshare==0.7.5,pluggy==1.0.0,prompt-toolkit==3.0.28,protobuf==3.19.4,ptyprocess==0.7.0,pure-eval==0.2.2,py==1.11.0,pyasn1==0.4.8,pyasn1-modules==0.2.8,pycparser==2.21,Pygments==2.11.2,pymacaroons==0.13.0,PyNaCl==1.5.0,pyparsing==3.0.7,pyRFC3339==1.1,pytest==7.0.1,pytest-asyncio==0.18.1,pytest-operator==0.11.0,python-dateutil==2.8.2,pytz==2021.3,PyYAML==6.0,requests==2.27.1,requests-oauthlib==1.3.1,rsa==4.8,six==1.16.0,stack-data==0.2.0,tenacity==8.0.1,theblues==0.5.2,toml==0.10.2,tomli==2.0.1,toposort==1.7,traitlets==5.1.1,typing-inspect==0.7.1,typing_extensions==4.1.1,urllib3==1.26.8,wcwidth==0.2.5,websocket-client==1.3.1,websockets==7.0
integration run-test-pre: PYTHONHASHSEED='1921150819'
integration run-test: commands[0] | pytest -v --tb native --ignore=/home/ubuntu/workdir/waltz-operator/tests/unit --log-cli-level=INFO -s
============================================================================ test session starts ============================================================================
platform linux -- Python 3.8.10, pytest-7.0.1, pluggy-1.0.0 -- /home/ubuntu/workdir/waltz-operator/.tox/integration/bin/python
cachedir: .tox/integration/.pytest_cache
rootdir: /home/ubuntu/workdir/waltz-operator
plugins: operator-0.11.0, asyncio-0.18.1
asyncio: mode=legacy
collected 6 items

tests/test_charm.py::TestCharm::test_config_changed
------------------------------------------------------------------------------- live log call -------------------------------------------------------------------------------
INFO     charm:charm.py:129 Waltz needs to be updated. Restarting.
PASSED
tests/test_charm.py::TestCharm::test_database_relation
------------------------------------------------------------------------------- live log call -------------------------------------------------------------------------------
INFO     charm:charm.py:129 Waltz needs to be updated. Restarting.
INFO     pgsql.client.db:client.py:368 emitting database_relation_joined event for relation 0
INFO     pgsql.client.db:client.py:402 emitting master_available event for relation 0
INFO     pgsql.client.db:client.py:407 emitting master_changed event for relation 0
INFO     charm:charm.py:129 Waltz needs to be updated. Restarting.
INFO     pgsql.client.db:client.py:432 emitting database_available event for relation 0
INFO     pgsql.client.db:client.py:436 emitting database_changed event for relation 0
INFO     pgsql.client.db:client.py:478 emitting master_changed event for relation 0
INFO     charm:charm.py:129 Waltz needs to be updated. Restarting.
INFO     pgsql.client.db:client.py:480 emitting master_gone event for relation 0
INFO     pgsql.client.db:client.py:490 emitting database_changed event for relation 0
INFO     pgsql.client.db:client.py:492 emitting database_gone event for relation 0
INFO     pgsql.client.db:client.py:495 emitting database_relation_broken event for relation 0
INFO     pgsql.client.db:client.py:499 cleaning up broken relation 0
PASSED
tests/test_charm.py::TestCharm::test_waltz_pebble_ready
------------------------------------------------------------------------------- live log call -------------------------------------------------------------------------------
INFO     charm:charm.py:129 Waltz needs to be updated. Restarting.
PASSED
tests/integration/test_charm.py::test_build_and_deploy
------------------------------------------------------------------------------ live log setup -------------------------------------------------------------------------------
INFO     pytest_operator.plugin:plugin.py:305 Using tmp_path: /home/ubuntu/workdir/waltz-operator/.tox/integration/tmp/pytest/test-charm-4h780
INFO     pytest_operator.plugin:plugin.py:385 Adding model legend-controller:test-charm-4h78
------------------------------------------------------------------------------- live log call -------------------------------------------------------------------------------
INFO     pytest_operator.plugin:plugin.py:489 Building charm finos-waltz-k8s
INFO     juju.model:model.py:1922 Deploying local:focal/finos-waltz-k8s-0
INFO     juju.model:model.py:1922 Deploying ch:amd64/focal/postgresql-k8s-3
INFO     juju.model:model.py:2694 Waiting for model:
  finos-waltz-k8s/0 [allocating] waiting: agent initializing
INFO     juju.model:model.py:2694 Waiting for model:
  finos-waltz-k8s/0 [idle] blocked: Waiting for database configuration.
INFO     juju.model:model.py:2694 Waiting for model:
  finos-waltz-k8s/0 [idle] blocked: Waiting for database configuration.
INFO     juju.model:model.py:2694 Waiting for model:
  finos-waltz-k8s/0 [idle] blocked: Waiting for database configuration.
INFO     juju.model:model.py:2694 Waiting for model:
  finos-waltz-k8s/0 [idle] blocked: Waiting for database configuration.
INFO     juju.model:model.py:2694 Waiting for model:
  finos-waltz-k8s/0 [idle] blocked: Waiting for database configuration.
INFO     juju.model:model.py:2694 Waiting for model:
  finos-waltz-k8s/0 [executing] active:
PASSED
tests/integration/test_charm.py::test_application_is_up
------------------------------------------------------------------------------- live log call -------------------------------------------------------------------------------
INFO     test_charm:test_charm.py:33 Trying to access Waltz...
INFO     test_charm:test_charm.py:39 HTTPConnectionPool(host='10.1.243.205', port=8080): Max retries exceeded with url: / (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7f4684493460>: Failed to establish a new connection: [Errno 111] Connection refused'))
INFO     test_charm:test_charm.py:33 Trying to access Waltz...
INFO     test_charm:test_charm.py:39 HTTPConnectionPool(host='10.1.243.205', port=8080): Max retries exceeded with url: / (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7f46886b47f0>: Failed to establish a new connection: [Errno 111] Connection refused'))
INFO     test_charm:test_charm.py:33 Trying to access Waltz...
INFO     test_charm:test_charm.py:39 HTTPConnectionPool(host='10.1.243.205', port=8080): Max retries exceeded with url: / (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7f46886c98e0>: Failed to establish a new connection: [Errno 111] Connection refused'))
INFO     test_charm:test_charm.py:33 Trying to access Waltz...
INFO     test_charm:test_charm.py:39 HTTPConnectionPool(host='10.1.243.205', port=8080): Max retries exceeded with url: / (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7f46844938b0>: Failed to establish a new connection: [Errno 111] Connection refused'))
INFO     test_charm:test_charm.py:33 Trying to access Waltz...
INFO     test_charm:test_charm.py:39 HTTPConnectionPool(host='10.1.243.205', port=8080): Max retries exceeded with url: / (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7f4683452ee0>: Failed to establish a new connection: [Errno 111] Connection refused'))
INFO     test_charm:test_charm.py:33 Trying to access Waltz...
INFO     test_charm:test_charm.py:39 HTTPConnectionPool(host='10.1.243.205', port=8080): Max retries exceeded with url: / (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7f468869b130>: Failed to establish a new connection: [Errno 111] Connection refused'))
INFO     test_charm:test_charm.py:33 Trying to access Waltz...
INFO     test_charm:test_charm.py:39 HTTPConnectionPool(host='10.1.243.205', port=8080): Max retries exceeded with url: / (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7f46886e2880>: Failed to establish a new connection: [Errno 111] Connection refused'))
INFO     test_charm:test_charm.py:33 Trying to access Waltz...
INFO     test_charm:test_charm.py:82 Successfully reached Waltz through its IP.
PASSED
tests/integration/test_charm.py::test_nginx_ingress_integration
------------------------------------------------------------------------------- live log call -------------------------------------------------------------------------------
INFO     juju.model:model.py:1922 Deploying ch:amd64/focal/nginx-ingress-integrator-25
INFO     juju.model:model.py:2694 Waiting for model:
  nginx-ingress-integrator/0 [allocating] waiting: installing agent
INFO     juju.model:model.py:2694 Waiting for model:
  nginx-ingress-integrator/0 [idle] waiting: agent initializing
INFO     juju.model:model.py:2694 Waiting for model:
  nginx-ingress-integrator/0 [executing] active: Ingress with service IP(s): 10.152.183.177
INFO     test_charm:test_charm.py:33 Trying to access Waltz...
INFO     test_charm:test_charm.py:110 Successfully reached Waltz through its service-hostname.
PASSED
----------------------------------------------------------------------------- live log teardown -----------------------------------------------------------------------------
INFO     pytest_operator.plugin:plugin.py:406 Model status:

Model            Controller         Cloud/Region        Version  SLA          Timestamp
test-charm-4h78  legend-controller  microk8s/localhost  2.9.25   unsupported  13:01:18Z

App                       Version                 Status  Scale  Charm                     Store     Channel  Rev  OS          Address         Message
finos-waltz-k8s                                   active      1  finos-waltz-k8s           local                0  kubernetes  10.152.183.61
nginx-ingress-integrator                          active      1  nginx-ingress-integrator  charmhub  stable    25  kubernetes  10.152.183.210
postgresql-k8s            .../postgresql@ed0e37f  active      1  postgresql-k8s            charmhub  stable     3  kubernetes

Unit                         Workload  Agent  Address       Ports     Message
finos-waltz-k8s/0*           active    idle   10.1.243.205
nginx-ingress-integrator/0*  active    idle   10.1.243.208            Ingress with service IP(s): 10.152.183.177
postgresql-k8s/0*            active    idle   10.1.243.207  5432/TCP  Pod configured


INFO     pytest_operator.plugin:plugin.py:412 Juju error logs:


INFO     pytest_operator.plugin:plugin.py:430 Destroying model test-charm-4h78


============================================================================= warnings summary ==============================================================================
.tox/integration/lib/python3.8/site-packages/pytest_asyncio/plugin.py:191
  /home/ubuntu/workdir/waltz-operator/.tox/integration/lib/python3.8/site-packages/pytest_asyncio/plugin.py:191: DeprecationWarning: The 'asyncio_mode' default value will change to 'strict' in future, please explicitly use 'asyncio_mode=strict' or 'asyncio_mode=auto' in pytest configuration file.
    config.issue_config_time_warning(LEGACY_MODE, stacklevel=2)

.tox/integration/lib/python3.8/site-packages/pytest_asyncio/plugin.py:226
  /home/ubuntu/workdir/waltz-operator/.tox/integration/lib/python3.8/site-packages/pytest_asyncio/plugin.py:226: DeprecationWarning: '@pytest.fixture' is applied to <fixture ops_test, file=/home/ubuntu/workdir/waltz-operator/.tox/integration/lib/python3.8/site-packages/pytest_operator/plugin.py, line=160> in 'legacy' mode, please replace it with '@pytest_asyncio.fixture' as a preparation for switching to 'strict' mode (or use 'auto' mode to seamlessly handle all these fixtures as asyncio-driven).
    warnings.warn(

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
================================================================= 6 passed, 2 warnings in 471.71s (0:07:51) =================================================================
__________________________________________________________________________________ summary __________________________________________________________________________________
  integration: commands succeeded
  congratulations :)
```

File: tests/integration/test_charm.py [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+    for i in range(30):
+        logger.info("Trying to access Waltz...")
+        try:
+            response = requests.get(url, headers=headers)
+            if response.status_code == 200:
+                return True
+        except Exception as ex:
+            logger.info(ex)
+
+        time.sleep(2)
```
Comment (jnsgruk) [inline tests/integration/test_charm.py:41]: This retry logic feels a bit naive, could we replace this with tenacity please? Something like [this](https://github.com/canonical/mongodb-operator/blob/5d5e7c1ba3b4a1a3a5c25d170bfc88c819386290/tests/integration/test_charm.py#L207)

**Review by jnsgruk — CHANGES_REQUESTED** [review]
One minor change on retry logic in integration tests.

File: tests/integration/test_charm.py (hunk 1 above)
Comment (claudiubelu) [reply tests/integration/test_charm.py]: Done

**Review by jnsgruk — APPROVED** [review]

**Pull request merged**